Add ISRC song search for YouTube

ehfd opened the pull request:
Fixes #883 and enhances Alternative Sources detection.

Reference:
https://isrc.ifpi.org/en/
https://www.isrcfinder.com/

Example:

![image](https://github.com/user-attachments/assets/f22c860b-e86a-4a9b-acb8-0ebffbb2a1c0)

CC @KRTirtho

KRTirtho commented:
That's a great addition. But does it work for all tracks, tho? Less popular track is what suffers most when matching

ehfd commented:
It gracefully falls back to the original implementation when the track isn't present. Tested that they work.

ehfd commented:
But it works when the track is present in YouTube Music.

KRTirtho commented:
We already use song.link for the exact task. If isrc doesn't match, that will only make extra requests to YouTube?
Btw with song.link we don't even have to search. It's like a database of already matched Spotify to YouTube, Soundcloud tracks

ehfd commented:
So:

1. This should provide an accurately matching track if the track is available on both Spotify and YouTube Music (it's surely a lot of tracks), which would definitely benefit many lesser-known tracks compared to song.link.
2. Even on well-known tracks, it points to the correct track that would be the same as Spotify, not the music video.

In short, this PR is meant to address:
![image](https://github.com/user-attachments/assets/5af4d524-23e4-4d30-b288-8cd25540a30c)

When the track is unavailable on YouTube Music but available on Spotify and there isn't a sensible search query, that is quite an extreme case, I'd say, which should be addressed separately.

ehfd commented:
@KRTirtho Provide me with the screenshot in the India region for the `DGA052354009` query.

KRTirtho commented:
Just search and you'll find nothing. Although the artist never released the track in YouTube in the first place. It's some other random channels that uploaded it there.
![image](https://github.com/user-attachments/assets/f9b64995-1da5-496a-b47c-97b58e803aed)

ehfd commented:
OK, that is a pretty extreme case. @KRTirtho 
I honestly think that this is a matter of a smarter query search from `rankResults` or a proper curation song.link, like how it's originally done.

But this PR is a totally different thing from song.link and greatly benefits tracks that are available on YouTube Music, whether they are more or less popular.

The filtering makes sure that the title of the query with the ISRC matches the title in Spotify.
It discards all other results and moves on to the existing methods.

ehfd commented:
One issue so far. @KRTirtho 

It seems that Track objects from albums are constructed differently from tracks in searches or artist pages.
Only tracks from albums don't have the `track.externalIDs.isrc` and are all `null`.

ehfd commented:
> Less popular track is what suffers most when matching

@KRTirtho My thought about this is: what if the prioritized metric for `rankResults` is the percentage of the YouTube Video's title matching the Spotify track's title and artist, after removing all of the special characters?

So, using your own example, use the string `Neshar Bojha - Popeye Bangladesh` for the query.
The query would return a series of YouTube videos, which we process the returned video titles with `.toLowerCase().replaceAll(RegExp(r'[^a-zA-Z0-9\s]+'), '').split(RegExp(r'\s+')).where((item) => item.isNotEmpty)`. This would lead to `['neshar', 'bojha', 'popeye', 'bangladesh']`.

The focus here below is not to use the scoring as the first-hand discriminator; instead, filter with the title, and then the author first.
Then, the initial priority (Score Number 1) is the percentage of which the video title is matched to Spotify's title. If a video title is `Neshar Bojha - Popeye Bangladesh`, this is 100% since `['neshar', 'bojha']` is contained. When the title is `Neshar Bojha`, it is also 100%. However, when the video title is only `Popeye Bangladesh`, it is 0%.

Then, this second priority (Score Number 2) is only utilized as a tiebreaker to the title score. That is, this is not combined as a score, but rather as a discriminator only when Score Number 1 is equal. This time, the Spotify title and the artist are both used, leading to ['neshar', 'bojha', 'popeye', 'bangladesh']. This time, `Neshar Bojha - Popeye Bangladesh`, with a score of 100%, is prioritized to `Neshar Bojha`, with a score of 50%.

The original scores such as official flag, artist name, etc. are only used as a tiebreaker to Score Number 3.

ehfd commented:
@KRTirtho 
So the characteristics of each method (in the order used):

1. ISRC search:
- Addresses the needs of users who listen to artists mainly located in North America, Europe, and East Asia, and have contracts with large record companies.
- 100% accurate if the track is licensed by YouTube (at least tens of millions of tracks).
- Contains no music videos; the track provided with this search is identical to Spotify, with correct sync to lyrics, solving a long-standing problem of Spotube that even Song.Link could not solve.

2. Song.Link
- User-curated links to YouTube.
- Relatively accurate if the track has been curated by users.
- Likely contains music videos; they mostly do not sync with lyrics.
- Errors exist for remixes, etc. - Curators may set the track to the original version even if the Spotify tack points to remixes.

3. Query Search
- Required for tracks that are not included in YouTube Music and Song.Link.
- Depends on the accuracy of the query scoring method.

ehfd commented:
@KRTirtho I rewrote all of the pull request and fixed all issues on my side.

There are subtle differences between Song.Link and ISRC streams; the ISRC stream being identical to Spotify.

ehfd commented:
Only skepticism I have remaining is whether in `lib/extensions/track.dart`:

```
    try {
      final spotify = ref.read(spotifyProvider);
      return await spotify.invoke((api) => api.tracks.get(id!));
    } catch (e, stack) {
      if ((e is AuthorizationException && e.error != 'invalid_token') ||
          e is SpotifyException) {
```

This is an adequate try-catch or not. I wrote it based on `lib/provider/spotify/spotify.dart`:

```
    try {
      return await fn(api);
    } catch (e) {
      if (((e is AuthorizationException && e.error == 'invalid_token') ||
              e is ExpirationException) &&
          !_isRefreshing) {
```

KRTirtho commented on lib/extensions/track.dart:
```diff
@@ -68,26 +70,34 @@ extension TrackExtensions on Track {
-    return track;
+  Future<Track> asTrack(AlbumSimple album, ref) async {
+    try {
+      final spotify = ref.read(spotifyProvider);
+      return await spotify.invoke((api) => api.tracks.get(id!));
```
This will make a request for each track of an album. Instead, just delete this extension method and replace it with [get-several-track](https://developer.spotify.com/documentation/web-api/reference/get-several-tracks) (`spotify.tracks.list(<ids>)` for dart)

Or create an extension method on `Iterable<TrackSimple>` where you map the ids of the `TrackSimple` then get the tracks using  `spotify.tracks.list(<ids>)`

`TrackSimple.asTrack` is used in 
- [lib/modules/album/album_card.dart](https://github.com/KRTirtho/spotube/blob/0ec9f3535bf9c927d81cb67e62aa35da36648c98/lib/modules/album/album_card.dart#L57)
- [lib/provider/spotify/album/tracks.dart](https://github.com/KRTirtho/spotube/blob/0ec9f3535bf9c927d81cb67e62aa35da36648c98/lib/provider/spotify/album/tracks.dart#L36)

KRTirtho commented on lib/modules/album/album_card.dart:
```diff
@@ -54,7 +54,8 @@ class AlbumCard extends HookConsumerWidget {
     Future<List<Track>> fetchAllTrack() async {
       if (album.tracks != null && album.tracks!.isNotEmpty) {
-        return album.tracks!.map((track) => track.asTrack(album)).toList();
+        return await Future.wait(
+            album.tracks!.map((track) => track.asTrack(album, ref)));
```
This will rate limit the user IP. Totally a no go

KRTirtho commented on lib/provider/spotify/album/tracks.dart:
```diff
@@ -33,7 +33,7 @@ class AlbumTracksNotifier extends AutoDisposeFamilyPaginatedAsyncNotifier<Track,
       (api) => api.albums.tracks(arg.id!).getPage(limit, offset),
     );
-    final items = tracks.items?.map((e) => e.asTrack(arg)).toList() ?? [];
+    final List<Track> items = await Future.wait(tracks.items?.map((e) => e.asTrack(arg, ref)) ?? []);
```
Again same issue

ehfd replied:
Resolved.

ehfd commented on lib/services/sourced_track/sources/youtube.dart:
```diff
@@ -240,25 +240,63 @@ class YoutubeSourcedTrack extends SourcedTrack {
 
-    if (ytLink?.url != null
-        // allows to fetch siblings more results for already sourced track
-        &&
-        track is! SourcedTrack) {
```
My understanding of these two lines is that it was valid when you used two separate `return` for Song.Link and the YouTube `title - artist` query.

In the implementation before this PR, during the initial sourcing, the function went into this `if` statement and returned right away when the Song.Link video is accessible.

It is only in the second call of this function after initial sourcing (or when Song.Link fails), that the YouTube `title - artist` query part of the code is reached.

However, I changed things around in the whole `fetchSiblings` function, where the YouTube `title - artist` query is done in the first initial sourcing as well, and combined with ISRC and/or Song.Link.

This is the reason the conversion `toSiblingType` is done last.

Therefore, I think that `track is! SourcedTrack` is inadequate in this context, and only skipping when there is an available ISRC track is adequate.

ehfd commented on lib/services/sourced_track/sources/youtube.dart:
```diff
@@ -240,25 +240,63 @@ class YoutubeSourcedTrack extends SourcedTrack {
+      }
+    }
+
     final links = await SongLinkService.links(track.id!);
     final ytLink = links.firstWhereOrNull((link) => link.platform == "youtube");
```
I accept this.

ehfd replied:
Resolved in https://github.com/KRTirtho/spotube/pull/2594/commits/8ad6c48bec146786e9ae67f9c7be59cb3fcc85be.

ehfd commented on lib/services/sourced_track/sources/youtube.dart:
```diff
@@ -271,20 +309,31 @@ class YoutubeSourcedTrack extends SourcedTrack {
+        }
+      }
+      return uniqueSiblings;
+    })();
```
What do you want me to do, in the end?

ehfd replied:
Note that I changed the syntax so that `return await Future.wait(videoResults.map((videoResult) async {...` is used instead.

ehfd commented:
@KRTirtho I have nothing more to touch.
Thanks for the review.

KRTirtho commented:
Ok, I'll test and see if everything works perfectly and merge onwards

KRTirtho commented:
Great work @ehfd
I tested all the core functionalities, and it's working great. Thanks for the contribution

ehfd commented:
@KRTirtho Great working with you, and it was a great detour from what I was typically doing.

ehfd commented:
Note that it seems that the issue with people skipping songs is because of `VideoUnplayableException`, and when this happens, the app goes to the next song instead of the next source.

Now that all sources are supplied within the first sourcing, this should be an easy solve for you.
